### PR TITLE
Sync Delegation

### LINKED ISSUE #88: $BB delelgation syncing
- [x] after minting $BB, a check to see if the holder is delegated to someone on the $BREAD contract should be performed , if they are delegated then after minting , a delegation to that holder should be performed instead of the self delegation 
- [x] A test should be added to confirm this behavior 
- [x] #90
- [x] The delegation entry point for the $BB contract should be disabled similarly to transfer 

This behavior does not guarantee 100% synced delegation behavior, as that would require the bread contract to call the buttered bread contract, which is an undesirable implementation pattern. This is a compromise, assuming that the user performs the LP deposit after delegation, or that partial withdraws \ new deposits will sync the delegation state.

### PR DESCRIPTION
**Pull request by daopunk**
closes #88

### CODE REVIEW

File: src/ButteredBread.sol [hunk 1]
```diff
@@ -115,7 +115,8 @@ contract ButteredBread is ERC20VotesUpgradeable, OwnableUpgradeable, IButteredBr
         _accountToLPData[_account][_lp].scalingFactor = currentScalingFactor;
 
         _mint(_account, _amount * currentScalingFactor / FIXED_POINT_PERCENT);
-        if (this.delegates(_account) == address(0)) _delegate(_account, _account);
+        if (this.delegates(_account) != address(0)) _delegate(_account, this.delegates(_account));
```
Comment (RonTuretzky) [inline src/ButteredBread.sol]: essentially here we need to check if in the $BREAD contract (forgot to mention that in the issue, we need to add that contract here) the user is delegated to someone. if they are delegated to someone, we need to delegate their $BB balance to that user as well. So if `if (BREAD.delegates(_account) != address(0)) _delegate(_account, BREAD.delegates(_account))`

**Review by RonTuretzky — CHANGES_REQUESTED** [review]

File: src/ButteredBread.sol [hunk 2]
```diff
@@ -115,14 +124,15 @@ contract ButteredBread is ERC20VotesUpgradeable, OwnableUpgradeable, IButteredBr
         _accountToLPData[_account][_lp].scalingFactor = currentScalingFactor;
 
         _mint(_account, _amount * currentScalingFactor / FIXED_POINT_PERCENT);
-        if (this.delegates(_account) == address(0)) _delegate(_account, _account);
+        _delegate(_account, BREAD.delegates(_account));
```
Comment (RonTuretzky) [inline src/ButteredBread.sol]:
```suggestion
       address delegatee = BREAD.delegates(_account);
       delegatee == address(0) ?  _delegate(_account,_account) : _delegate(_account,delegatee ) ;
```

File: src/ButteredBread.sol [hunk 3]
```diff
@@ -115,14 +124,15 @@ contract ButteredBread is ERC20VotesUpgradeable, OwnableUpgradeable, IButteredBr
...
 
         emit AddButter(_account, _lp, _amount);
     }
 
     /// @notice Withdraw LP tokens and burn ButteredBread with corresponding LP scaling factor
     function _withdraw(address _account, address _lp, uint256 _amount) internal {
         if (_amount > _accountToLPData[_account][_lp].balance) revert InsufficientFunds();
+        _delegate(_account, BREAD.delegates(_account));
```
Comment (RonTuretzky) [inline src/ButteredBread.sol]:
```suggestion
        address delegatee = BREAD.delegates(_account);
        delegatee == address(0) ?  _delegate(_account,_account) : _delegate(_account,delegatee ) ;
```

File: src/interfaces/IERC20Votes.sol [hunk 4]
```diff
@@ -0,0 +1,44 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity ^0.8.25;
+
+import {IERC20} from "openzeppelin-contracts/contracts/token/ERC20/IERC20.sol";
```
Comment (RonTuretzky) [inline src/interfaces/IERC20Votes.sol:4]: really silly that oz does not provide an interface for this class , but I see no better option than what you did 

thoughts @bagelface ?

**Review by RonTuretzky — CHANGES_REQUESTED** [review]

File: src/ButteredBread.sol (hunk 2 above)
Comment (RonTuretzky) [reply src/ButteredBread.sol]: this snippet could also instead be made into a function that anyone can call , would reduce the amount of code and also allow a bypass for mismatched state

File: src/interfaces/IERC20Votes.sol (hunk 4 above)
Comment (bagelface) [reply src/interfaces/IERC20Votes.sol:4]: seems fine to me

File: src/interfaces/IButteredBread.sol [hunk 5]
```diff
@@ -11,6 +11,8 @@ interface IButteredBread {
     error InvalidValue();
     /// @notice Occurs when attempting a deposit with a non-sanctioned LP
     error NotAllowListed();
+    /// @notice Occurs when attempting to delegate ButteredBread that is determined by BREAD
```
Comment (bagelface) [inline src/interfaces/IButteredBread.sol]: I was sort of confused when reading this comment. Would suggest making it more clear. Suggestion:
`Occurs when attempting to delegate ButteredBread. Delegations are set via the BREAD contract`

File: test/ButteredBread.t.sol [hunk 6]
```diff
@@ -403,3 +406,60 @@ contract ButteredBreadTest_Fuzz is ButteredBreadTest {
...
+        assertEq(bb.delegates(BOBBY), DELEGATEE);
+    }
+
+    function testDelegationRevert() public {
+        vm.startPrank(ALICE);
+        bb.deposit(GNOSIS_CURVE_POOL_XDAI_BREAD, TOKEN_AMOUNT / 3);
+
+        assertEq(bb.delegates(ALICE), ALICE);
+
+        vm.expectRevert();
```
Comment (bagelface) [inline test/ButteredBread.t.sol]: prefer testing for explicit errors
`vm.expectRevert(abi.encodeWithSelector(IButteredBread.NonDelegatable.selector));`

File: src/ButteredBread.sol (hunk 2 above)
Comment (bagelface) [reply src/ButteredBread.sol]: seems reasonable. also can just call the function code in `_deposit` `_withdraw` instead of repeating code

**Review by bagelface — APPROVED** [review]

File: test/ButteredBread.t.sol (hunk 6 above)
Comment (daopunk) [reply test/ButteredBread.t.sol]: oh wow, thank you.... i have had so much trouble formatting custom errors in revert statements, which is why i didn't do it. glad to see the correct format

**Review by RonTuretzky — APPROVED** [review]

**Pull request merged**